Supported FUNIKI remote AC

QuangThai2297 opened the pull request:
This merge request adds basic support for the FUNIKI remote. This includes support for almostly decode data from remote as power on/off, temperature, mode, fan, clock, timer on/off, swingV, sleep,.. and support sending almostly setting except timer on/off. 
For information about this remote, you can refer this issue at https://github.com/crankyoldgit/IRremoteESP8266/issues/2112
It is old remote model. Data has no checksum byte, and is very easy to decode.
Thanks!

QuangThai2297 committed:
supported FUNIKI remote AC

crankyoldgit committed:
Update IRac.h

Change argument order and remove unused parameters.

crankyoldgit committed:
Update IRac.cpp

Change argument order and remove unused parameters.
Fix a doxygen error
Fix a typo/error in the Ptr comment.

crankyoldgit commented on src/ir_Funiki.h:
```diff
@@ -0,0 +1,204 @@
+/// @brief Support for FUNIKI A/C protocols.
+/// @see https://github.com/crankyoldgit/IRremoteESP8266/issues/2112
+
+
```
Please add the supported devices info here please. It's required for our automatic documentation.
e.g. Like the following for Kelvinator.
https://github.com/crankyoldgit/IRremoteESP8266/blob/dc0fd31706ff26fbd28b004a971ea4a702fb5e60/src/ir_Kelvinator.h#L5-L19

QuangThai2297 replied:
Thank you! Let me check!

QuangThai2297 commented:
Can anyone help me resolve this failing after check PRs below?
"Library Linter / supported-devices-check (pull_request) Failing after 6s"
I am new here and this is the first time I contribute this project! Thanks!

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+  (void)(length);
+  // Funiki uses the same checksum alg. as Kelvinator's block checksum.
+  // _.Sum = IRKelvinatorAC::calcBlockChecksum(_.remote_state, length);
+}
```
This doesn't make sense. If there is no checksum, then you shouldn't need this procedure.
And the `validChecksum()` function shouldn't be needed. If it is needed, the the checksum should do something.

If they are both un-needed, they should be removed.

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+
+/// Get/Detect the model of the A/C.
+/// @return The enum of the compatible model.
+funiki_ac_remote_model_t IRFunikiAC::getModel(void) const { return _model; }
```
This seems like it doesn't do anything. If the model functions are not needed, they should be removed.

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+//   _.TimerTensHr = hours / 10;
+//   // Set the "units" digit of hours.
+//   _.TimerHours = hours % 10;
+// }
```
Please remove the commented out function. Try to avoid including code that doesn't work etc. Someone might think it works and try it.

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+stdAc::state_t IRFunikiAC::toCommon(void) {
+  stdAc::state_t result{};
+  result.protocol = decode_type_t::FUNIKI;
+  result.model = _model;
```
If `model`s isn't used, return -1 (I think)

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+String IRFunikiAC::toString(void) {
+  String result = "";
+  result.reserve(220);  // Reserve some heap for the string to reduce fragging.
+  result += addModelToString(decode_type_t::FUNIKI, _model, false);
```
Do we even need Model stuff?

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+  // for(int i = 0; i < results->rawlen; i++)
+  // {
+  //   printf("%d, ",results->rawbuf[i]);
+  // }
```
Remove these comments please.

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+  if (strict && nbits != kFunikiBits)
+    return false;  // Not strictly a Funiki message.
+
+  // printf("\nlen: %d\n", results->rawlen);
```
Remove commented code.

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+  // if (strict) {
+  //   // Verify the message's checksum is correct.
+  //   if (!IRFunikiAC::validChecksum(results->state)) return false;
+  // }
```
Remove commented out code please.

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+                      tole, kMarkExcess, false);
+  if (used == 0) return false;
+  offset += used;
+  // printf("\n ooffset: %d\n", offset);
```
Remove commented out code please.

crankyoldgit commented on src/ir_Funiki.cpp:
```diff
@@ -0,0 +1,580 @@
+                          kFunikiBitMark, kFunikiOneSpace, kFunikiBitMark,
+                          kFunikiZeroSpace, tole, kMarkExcess, false);
+  if (data_result.success == false) return false;
+  // printf("\n foot: %lld\n", data_result.data);
```
Remove commented out code please.

QuangThai2297 replied:
Surely. I will do.

crankyoldgit reviewed (commented):
Thanks for this. Looks mostly good.
Can you supply some raw data and  what the "raw" A/C message is supposed to be/do etc. i.e. the settings state.

crankyoldgit commented:
> Can anyone help me resolve this failing after check PRs below? "Library Linter / supported-devices-check (pull_request) Failing after 6s" I am new here and this is the first time I contribute this project! Thanks!

I've given you a code review. See the review comment for the `ir_Funiki.h` file.

Basically you are missing some info in the header file that we scan for to help automatically build some of the documentation for the project. Namely `https://github.com/crankyoldgit/IRremoteESP8266/blob/master/SupportedProtocols.md`

QuangThai2297 committed:
clean and remove the unused

QuangThai2297 commented:
> Thanks for this. Looks mostly good. Can you supply some raw data and what the "raw" A/C message is supposed to be/do etc. i.e. the settings state.

Now I have just updated new commit. I removed un-needed something as you recommended. And you can also see "All checks have passed". 
For raw message from remote capture, you can refer attached file below:
[FUNIKI_remote_AC_Raw_code_decode.txt](https://github.com/user-attachments/files/16045245/FUNIKI_remote_AC_Raw_code_decode.txt)
